Add a flag to control whether DocumentSnapshots return Dates or Timestamps for timestamp fields

var-const opened the pull request.

var-const commented on Firestore/Source/API/FIRDocumentSnapshot.mm:
```diff
@@ -164,7 +167,8 @@ - (nullable id)valueForField:(id)field options:(FIRSnapshotOptions *)options {
              : [self convertedValue:fieldValue
-                            options:[FSTFieldValueOptions optionsForSnapshotOptions:options]];
+                            options:[FSTFieldValueOptions optionsForSnapshotOptions:options
+                            timestampBehavior:[self.firestore.settings timestampBehavior]]];
```
Passing timestampBehavior seems better than passing the whole settings object?

var-const commented on Firestore/Source/Model/FSTFieldValue.h:
```diff
@@ -17,6 +17,7 @@
 #import <Foundation/Foundation.h>
 
 #import "Firestore/third_party/Immutable/FSTImmutableSortedDictionary.h"
+#import "FIRTimestampBehavior.h"
```
I left `FSTFieldValueOptions` pretty much as it were.

var-const commented on Firestore/Source/Model/FSTFieldValue.mm:
```diff
@@ -47,28 +47,47 @@
-+ (instancetype)optionsForSnapshotOptions:(FIRSnapshotOptions *)options {
-  if (options.serverTimestampBehavior == FSTServerTimestampBehaviorNone) {
-    static FSTFieldValueOptions *defaultInstance = nil;
-    static dispatch_once_t onceToken;
```
Since you mentioned in Android review that default options seem unnecessary, perhaps I can delete them?

var-const replied:
@wilhuff, please let me know if you think this should be kept.

wilhuff replied:
Doesn't matter much to me. Doing it the old way saves a heap allocation.

Once tokens are going away anyway though so I wouldn't stress about this:
  * Most of our types are migrating to value types
  * Simple value types don't require a heap allocation so we wouldn't be saving anything necessarily by sharing
  * Magic statics make this work otherwise if there's benefit in sharing

var-const commented on Firestore/Source/Model/FSTFieldValue.mm:
```diff
@@ -47,28 +47,47 @@
-    return defaultInstance;
-  } else {
-    return [[FSTFieldValueOptions alloc]
-        initWithServerTimestampBehavior:options.serverTimestampBehavior];
```
It looks like the `FST` version of the enum was initialized by the `FIR` version here. I don't know if it's intentional/valid in Objective C.

var-const commented on Firestore/Source/Model/FSTFieldValue.mm:
```diff
@@ -47,28 +47,47 @@
+      convertedTimestampBehavior = FSTTimestampBehaviorReturnTimestamp;
+      break;
+    default:
+      FSTFail(@"Unexpected server timestamp option: %d", (int)timestampBehavior);
```
TODO: copy-paste error.

var-const replied:
Obsolete.

var-const commented on Firestore/Source/Model/FSTFieldValue.mm:
```diff
@@ -89,7 +108,8 @@ - (FSTTypeOrder)typeOrder {
   return [self valueWithOptions:[FSTFieldValueOptions
-                                    optionsForSnapshotOptions:[FIRSnapshotOptions defaultOptions]]];
+                                    optionsForSnapshotOptions:[FIRSnapshotOptions defaultOptions]
+                                    timestampBehavior:FIRTimestampBehaviorDefault]];
```
I really don't like this. It looks like this is the place in code where the default is actually set, regardless of what's inside `FIRFirestoreSettings`. I'm trying to work around having two places where default is set by introducing an "alias" for the default enum value. Let me know what you think.

mikelehen replied:
Am I correct in thinking that this code is only hit by tests (FIRDocumentSnapshot always uses the valueWithOptions: method instead)?  If so, then it seems like we should either:

1. Remove the method and update all tests to use valueWithOptions.
2. Use FIRTimestampBehaviorReturnTimestamp and update our tests to deal with it (since this will be the default in the future).
3. Leave this as FIRTimestampBehaviorReturnDate but put a comment explaining this is only used by tests and we should consider updating them to work with the new timestamp behavior.

mikelehen replied:
I think that would be an improvement, and I'd be fine with that approach.

Unfortunately I think FSTFieldValue is in a bit of a wonky state, especially on iOS. Note the comment on the `value` method:
```
 * Converts an FSTFieldValue into the value that users will see in document snapshots.
 *
 * TODO(mikelehen): This conversion should probably happen at the API level and right now `value` is
 * used inappropriately in the serializer implementation, etc.  We need to do some reworking.
 ```

Since `value` is meant to compute the _user-facing_ value, we shouldn't actually be using it in FSTRelationFilter or FSTSerializerBeta and expecting to get back a specific internal representation...

Instead we should directly expose the internal representation, e.g. in FSTFieldValue.h, add a property to FSTFieldValue:
```
@interface FSTReferenceValue : FSTFieldValue <FSTDocumentKey *>
...
@property(nonatomic, strong, readonly) FSTDocumentKey *internalValue;
@end
```

Several other FSTFieldValue subclasses already do this (search for internalValue in FSTFieldValue.h).  And then we could use `internalValue` from `FSTRelationFilter` instead of `value`. Applying the same rework to FSTSerializerBeta (and the tests?) we could ultimately get rid of the `value` method entirely. Internal code should use `internalValue` and the API layer code should use `valueWithOptions`.

And in the longer-term, I think we should actually get rid of `valueWithOptions` too, moving the "model types" => "API types" conversion in the API layer instead, which would clean up some other warts (like the fact that FSTReferencValue's `value*` method returns an FSTDocumentKey which then must be converted again into a FIRDocumentReference at the API layer).

All of this said, I know Zhi has already done FieldValue C++ porting work and I don't know how all of this overlaps... and we probably don't want to turn your PR into a "FieldValue revamp" PR. :-)  So feel free to pick whatever feels appropriate for your PR.  As I said at the beginning, your suggestion to flip the value / valueWithOptions relationship seems fine for now.

var-const replied:
I agree, and I was thinking along similar lines:
* `value` and `valueWithOptions` seem to be pretty different despite the name similarity - `value` is some sort of a "canonical" representation, and `valueWithOptions` is "user-facing";
* `DocumentSnapshot` might be a better place to apply the options. It will require type casting which isn't perfect, but having all `FieldValue`s implement `valueWithOptions` when in fact nearly all of them ignore options also isn't. I can work on that after Timestamp.

For the purposes of this PR, I did reverse the forwarding between `value` and `valueWithOptions`, and make `FSTSerializerBeta` call `[timestampValue value]` instead of `[(FSTTimestampValue*)timestampValue internalValue]`.

var-const commented on Firestore/Source/Public/FIRTimestampBehavior.h:
```diff
@@ -0,0 +1,38 @@
+/*
```
I had to split out this enum. Otherwise, I ran into a situation where `FSTFieldValue.h` would have to import `FIRFirestoreSettings.h`, and XCode started complaining about everything inside `FIRFirestoreSettings.h` being redefined. I'm not sure how Objective C deals with double imports, perhaps my "fix" is wrong.

var-const replied:
Because it's now just a boolean flag, this is no longer an issue.

var-const commented on Firestore/Source/Public/FIRTimestampBehavior.h:
```diff
@@ -0,0 +1,38 @@
+NS_ASSUME_NONNULL_BEGIN
+
+/**
+ * OBC
```
Just in case, "OBC" is a placeholder I use for myself (like a TODO, except I don't have to browse through everyone's TODOs).

mikelehen commented on Firestore/Source/Public/FIRTimestampBehavior.h:
```diff
@@ -0,0 +1,38 @@
+   */
+  FIRTimestampBehaviorReturnNativeDate,
+
+  FIRTimestampBehaviorDefault = FIRTimestampBehaviorReturnNativeDate
```
I'm not sure we actually want to expose FIRTimestampBehaviorDefault in our public API since its meaning will change in the future...

On that note, maybe we should just drop the enum and have:

settings.enableNewTimestampBehavior = true

And we'll log a warning if they don't set it to true...  and then we'll deprecate it once it's the default, and then delete it eventually...

var-const commented on Firestore/Source/Model/FSTFieldValue.mm:
```diff
@@ -47,28 +47,47 @@
+      convertedServerTimestampBehavior = FSTServerTimestampBehaviorPrevious;
+      break;
+    default:
+      FSTFail(@"Unexpected server timestamp option: %d", (int)options.serverTimestampBehavior);
```
Done.

var-const commented on Firestore/Source/Model/FSTFieldValue.mm:
```diff
@@ -907,14 +931,22 @@ - (NSUInteger)hash {
+- (id)value {
   NSMutableArray *result = [NSMutableArray arrayWithCapacity:_internalValue.count];
   [self.internalValue enumerateObjectsUsingBlock:^(FSTFieldValue *obj, NSUInteger idx, BOOL *stop) {
     [result addObject:[obj value]];
```
I'll write a test for this, but it was probably a bug (`options` weren't propagated to the elements in the array). Unless perhaps it's impossible to have `ServerTimestampValue`s in an array?

mikelehen replied:
It is indeed impossible to have server timestamp values in an array [because the protocol currently doesn't define how to reference an array element via a field path, so we can't create the necessary FieldTransform], but that should be rejected in the API layer (in FSTUserDataConverter probably)...  so it was probably a bug, although benign.

var-const replied:
Sounds like it would be hard/impossible to test?

mikelehen commented on Firestore/Source/Model/FSTFieldValue.h:
```diff
@@ -194,15 +198,15 @@ typedef NS_ENUM(NSInteger, FSTServerTimestampBehavior) {
 @interface FSTGeoPointValue : FSTFieldValue <FIRGeoPoint *>
 + (instancetype)geoPointValue:(FIRGeoPoint *)value;
-- (FIRGeoPoint *)valueWithOptions:(FSTFieldValueOptions *)options;
+- (FIRGeoPoint *)value;
```
So I'm having trouble figuring out why we were declaring valueWithOptions before and why you've changed it to declare value now...  Both are declared on the superclass (FSTFieldValue) as returning T and so there should be no need to declare them here, right?  Or am I confused?

FWIW (no action required)- in JS / Android FieldValue is not generic (I was surprised to see that it is in iOS), in which case it makes a little bit of sense for each subclass to declare the `value` methods so that they can refine the return type... except per my previous comments I think ultimately we should only be using `value` / `valueWithOptions` to convert to user-facing types and therefore we shouldn't need a return type more specific than `id`...

var-const replied:
I removed these declarations, and the test suite still runs fine. These might be leftovers: it looks like generics were added pretty recently in [this commit](https://github.com/firebase/firebase-ios-sdk/commit/64066c643d4e1fff7d00d656201d056f10fb2c36), and some of these `valueWithOptions` were removed, but some were kept. My guess is that all of them should have been removed.

mikelehen commented on Firestore/Source/Model/FSTFieldValue.mm:
```diff
@@ -497,6 +509,10 @@ - (FSTTypeOrder)typeOrder {
 }
 
+- (id)value {
+  return self.localWriteTime;
```
Hrm...  This makes me nervous because the default behavior is FSTServerTimestampBehaviorNone which returns NSNull...  So it seems like this should return NSNull or have a comment explaining what's going on.

var-const replied:
Hmm, I'm not sure. My understanding was that `value` should return some sort of a canonical representation of the object, which is independent from user-side settings (including the default behavior). So it seemed that `NSNull` is less appropriate - it seems to be just a way to drop the value. Another downside, in my view, is that the default behavior is now specified in two different places.

I don't feel strongly, though - feel free to push back.

mikelehen replied:
Honestly `value` shouldn't exist and I'm assuming this code isn't exercised since (I believe) you changed the value and it didn't break anything. And I'm not sure there is a good "canonical representation" for a server timestamp.  So I'd rather continue returning NSNull.

Keep in mind that on Android / JS, value and valueWithOptions are overloads of a single method, so I think it'd be strange to say that if you call it with no parameters it's the "canonical representation" but if you call it with an options param then it's the "user representation."

var-const replied:
Done.

mikelehen commented on Firestore/Source/Model/FSTFieldValue.mm:
```diff
@@ -506,7 +522,7 @@ - (id)valueWithOptions:(FSTFieldValueOptions *)options {
       return self.previousValue ? [self.previousValue valueWithOptions:options] : [NSNull null];
     default:
-      FSTFail(@"Unexpected server timestamp option: %d", (int)options.serverTimestampBehavior);
+      FSTFail(@"Unexpected server timestamp option: %ld", (long)options.serverTimestampBehavior);
```
Thanks. :-)

var-const commented on Firestore/Source/API/FIRDocumentSnapshot.mm:
```diff
@@ -141,8 +143,12 @@ - (FIRSnapshotMetadata *)metadata {
+                           options:[FSTFieldValueOptions
+                                         optionsForSnapshotOptions:options
+                                       enableTimestampsInSnapshots:
+                                           [self.firestore.settings enableTimestampsInSnapshots]]];
```
Done.

var-const commented on Firestore/Source/API/FIRFirestore.mm:
```diff
@@ -217,6 +217,11 @@ - (void)setSettings:(FIRFirestoreSettings *)settings {
                             "longer be changed. You can only set settings before calling any "
                             "other methods on a Firestore instance.");
     }
+    if (!settings.enableTimestampsInSnapshots) {
```
Any suggestions about the text are welcome. Also, I'm not sure if this warrants `FSTWarn`, or if it should be `FSTLog`.

mikelehen replied:
I think it needs to be FSTWarn to be visible-by-default, so I think this is right... And we may want to double-check that it really is visible-by-default when you run your app from xcode during development. The icore logging infrastructure is a bit nuanced I think...

Also, I think we probably want to move this check / log message to where we create FSTFirestoreClient (i.e. when the settings are used), since we want to log this message in the case they don't explicitly set the settings and just use the default ones.

var-const replied:
Done.

var-const commented on Firestore/Source/Public/FIRFirestoreSettings.h:
```diff
@@ -44,6 +44,18 @@ NS_SWIFT_NAME(FirestoreSettings)
+ * logged if enableTimestampsInSnapshots is set to false. Please upgrade your code to expect
+ * FIRTimestamp instead of NSDate and set enableTimestampsInSnapshots to true.
+ */
+@property(nonatomic) BOOL enableTimestampsInSnapshots;
```
Any suggestions about the wording are welcome.

wilhuff replied:
Here's my suggestion:

> Enables the use of FIRTimestamps for timestamp fields in FIRDocumentSnapshots.
> 
> Currently, Firestore returns timestamp fields as an `NSDate` but `NSDate` is implemented as a `double` which loses precision and causes unexpected behavior when using a timestamp from a snapshot as a part of a subsequent query.
>
> Setting `enableTimestampsInSnapshots` to `true` will cause Firestore to return `FIRTimestamp` values instead of `NSDate`, avoiding this kind of problem. To make this work you must also change any code that uses `NSDate` to use `FIRTimestamp` instead.
>
> NOTE: in the future `enableTimestampsInSnapshots = true` will become the default and this option will be removed so you should change your code to use FIRTimestamp now and opt-in to this new behavior as soon as you can.

mikelehen replied:
nit: `true` should be `YES` if we're using Obj-C syntax (which we seem to be with `FIRTimestamp` etc.)

var-const replied:
Gil, changed to use your wording.

Michael, I was originally following the comment on `persistenceEnabled`, which used `false`. I don't mind either way, but I think this should be consistent. Let's change both?

mikelehen replied:
Ahh, I see. My preference would be to ultimately move towards Swift syntax since that's the dominant language of our users (which would mean changing `FIRTimestamp` to `Timestamp` or maybe "Firebase `Timestamp`" in this case since "Timestamp" could be ambiguous on its own).

But I don't think we should try to tackle that in this PR. :-) So I think leaving this as-is is fine.

mikelehen commented on Firestore/Source/API/FIRFirestore.mm:
```diff
@@ -217,6 +217,13 @@ - (void)setSettings:(FIRFirestoreSettings *)settings {
+          @"FirestoreSettings.enableTimestampsInSnapshots is set to false. This is legacy "
+           "behavior which will be removed in a future release. Please set "
+           "enableTimestampsInSnapshots to true and upgrade your code to support the new "
+           "behavior.");
```
I'd phrase this more strongly and make it clearer what actions they need to take...  Something like:

```
The default behavior for NSDate objects stored in Firestore is going to change
AND YOUR APP MAY BREAK.

To hide this warning and ensure your app does not break, you need to add
the following code to your app before calling any other Cloud Firestore methods:

let db = Firestore.firestore()
let settings = db.settings
settings.enableTimestampsInSnapshots = true
db.settings = settings

With this change, timestamps stored in Cloud Firestore will be read back as
Firebase Timestamp objects instead of as system Date objects. So you will 
also need to update code expecting a Date to instead expect a Timestamp. 
For example:

// old: 
let date: Date = documentSnapshot.data["created_at"]
// new:
let timestamp: Timestamp = documentSnapshot.data["created_at"]
let date: Date = timestamp.approximateDateValue()

Please audit all existing usages of Date when you enable the new behavior. This new
behavior will become the default in a future release and so if you do not follow these
steps, YOUR APP MAY BREAK.
```

FYI- I didn't test my code snippets and my Swift is shaky, so they need validating. :-)

Alternatively, we could perhaps put this information in a doc page somewhere and link to it from this log message. This would give us the chance to improve / fix it if necessary, but also may mean it's less likely to be noticed and addressed. @samtstern, any thoughts?

Whatever we end up with here needs to be included in our official API Review, as on-by-default log messages are considered part of our API (and I think we want feedback on our breaking change migration strategy anyway).

samtstern replied:
Strong warning in-place is always nice, but you can also add a goo.gl link to docs at the bottom/top that can point to a docs page. The reason for the short link is so that we can change where it goes if we move docs around.

Docs page has the advantage of knowing the current version, timelines, etc.

mikelehen replied:
Thanks Sam...  Is there a sane place to put this sort of temporary migration information in the docs?  Or would it basically be a dangling page that's only linked from this warning message (and probably release notes, etc.)?

samtstern replied:
I think release notes is the right place.

var-const replied:
Michael, I added your suggestion as is for now, assuming it's non-final due to pending API review.

var-const commented on Firestore/Example/Tests/Integration/API/FIRFieldsTests.mm:
```diff
@@ -221,28 +225,77 @@ - (void)testFieldsWithSpecialCharsCanBeUsedInOrderBy {
+
+// Settings cannot be redefined once passed to Firestore instance, which necessitates overriding
+// the base class method to create different settings.
+// TODO(b/73820332): there will be no need for this once enableTimestampsInSnapshots is the default.
```
Moved the comment.

Re. `setUp`: I haven't tried it, but it seems like it won't work: FIRFirestore [won't allow](https://github.com/firebase/firebase-ios-sdk/blob/master/Firestore/Source/API/FIRFirestore.mm#L212-L217) redefining settings. Am I missing something?

mikelehen replied:
It won't allow it after _client is initialized, but _client isn't initialized until you call some method on FIRFirestore, and I /think/ the existing setUp method does not...  so I /think/ you can still modify settings in a setUp method here.

var-const replied:
Yes it works, thanks for explaining.

var-const commented on Firestore/Example/Tests/Integration/API/FIRFieldsTests.mm:
```diff
@@ -221,28 +225,77 @@ - (void)testFieldsWithSpecialCharsCanBeUsedInOrderBy {
+
+  FIRDocumentSnapshot *snapshot = [self readDocumentForRef:doc];
+  NSDictionary<NSString *, id> *data = [snapshot data];
+  // Timestamp must be truncated to microseconds after being written to the database.
```
Done.

mikelehen commented on Firestore/Source/API/FIRFirestore.mm:
```diff
@@ -243,6 +236,34 @@ - (void)ensureClientConfigured {
+             "let date: Date = timestamp.approximateDateValue()"
+             "\n\n"
+             "Please audit all existing usages of Date when you enable the new behavior. This new"
+             "behavior will become the default in a future release and so if you do not follow "
```
trailing space

var-const replied:
Unless Objective C concatenates string literals differently from C, it was actually the other way round - other lines were missing spaces, so that the last word from the previous line and the first word of the current line will be lumped together. Fixed.

mikelehen replied:
Oh, right! Good call.

mikelehen commented on Firestore/Source/API/FIRFirestore.mm:
```diff
@@ -243,6 +236,34 @@ - (void)ensureClientConfigured {
+             "\n\n"
+             "Please audit all existing usages of Date when you enable the new behavior. This new"
+             "behavior will become the default in a future release and so if you do not follow "
+             "these  steps, YOUR APP MAY BREAK.");
```
extra space

var-const replied:
Done.

mikelehen commented on Firestore/Source/API/FIRFirestore.mm:
```diff
@@ -243,6 +236,34 @@ - (void)ensureClientConfigured {
       FSTAssert(_settings.host, @"FirestoreSettings.host cannot be nil.");
       FSTAssert(_settings.dispatchQueue, @"FirestoreSettings.dispatchQueue cannot be nil.");
 
+      if (!_settings.enableTimestampsInSnapshots) {
```
Can you add something like:

// TODO(var-const): DO NOT SUBMIT: Double-check these code snippets before submitting.

just to make sure we don't forget? :)  I don't want to be responsible for throwing a warning in every developer's faces with code that doesn't actually work. :-)

var-const replied:
Done.

mikelehen commented on Firestore/Source/API/FIRFirestore.mm:
```diff
@@ -236,6 +236,35 @@ - (void)ensureClientConfigured {
+      if (!_settings.enableTimestampsInSnapshots) {
+        // TODO(varconst): DO NOT SUBMIT: doublecheck these code snippets before submitting.
+        FSTWarn(
+            @"The default behavior for NSDate objects stored in Firestore is going to change "
```
FYI- `NSDate` is Obj-C specific while the rest of this warning is Swift-oriented. I'd be inclined to say "timestamps" (which can be read generically or as the actual [Firestore term](https://developers.google.com/protocol-buffers/docs/reference/google.protobuf#google.protobuf.Timestamp)) or "system Date objects" (which is my best attempt to reference [NSDate using Swift terminology](https://developer.apple.com/documentation/foundation/date), but feel free to come up with something better).

Feel free to leave the PR as-is for now, but consider incorporating this along with final API review feedback, etc.

var-const commented:
@mikelehen Michael, could you please take another look? Changes since last time:
* tests use new behavior by default (like other platforms). Sorry, I had this on my laptop for a long time, but forgot to push
* slight tweaking of warning due to feedback on the doc. The idea was to phrase it clearer that the current behavior is *not* going to be supported in the future. I'll be happy to change if you'd like.

For testing, I built FireEats with this a couple of weeks ago and made sure that both code snippets work when copy-pasted into the app.

var-const commented:
Oh, sorry, this is iOS, so perhaps I should have assigned it to @wilhuff Gil... Anyway, I forgot that this PR was never explicitly approved, so I would need that before I can merge.

mikelehen commented on Firestore/Example/Tests/Integration/API/FIRFieldsTests.mm:
```diff
@@ -221,28 +225,83 @@ - (void)testFieldsWithSpecialCharsCanBeUsedInOrderBy {
-  NSDate *resultNestedDate = result[@"metadata.nestedTimestamp"];
-  XCTAssertEqualWithAccuracy([resultNestedDate timeIntervalSince1970], [date timeIntervalSince1970],
-                             0.000001);
+// Settings can only be redefined before client is initialized, so this has to happen in setUp.
```
This comment seems kind of out of place. If we have it at all, it should perhaps be in the setUp method? :-)

var-const replied:
Done.

mikelehen commented on Firestore/Source/API/FIRFirestore.mm:
```diff
@@ -242,6 +242,35 @@ - (void)ensureClientConfigured {
       FSTAssert(_settings.dispatchQueue, @"FirestoreSettings.dispatchQueue cannot be nil.");
 
+      if (!_settings.timestampsInSnapshotsEnabled) {
+        // TODO(varconst): DO NOT SUBMIT: doublecheck these code snippets before submitting.
```
Remove now?

var-const replied:
Done.

mikelehen commented:
(And I don't think @wilhuff needs to look unless he wants to.)

mikelehen commented:
Oh! Can you add an entry to Firestore/CHANGELOG.md for this too (in JS too, please)?

mikelehen commented on Firestore/CHANGELOG.md:
```diff
@@ -2,6 +2,16 @@
+  currently supporting up to microsecond precision. It can be passed to API
+  methods anywhere a system Date is currently accepted. To make
+  `DocumentSnapshot`s read timestamp fields back as `Timestamp`s instead of
+  Dates, set the newly-added flag `areTimestampsInSnapshotsEnabled` to `true` in
```
Optional suggestions:
* "set" => "you can set"
* move "to true" after "in FirestoreSettings"

var-const replied:
Done, thanks!

var-const commented:
@mikelehen, is there also a relevant changelog on Android?

wilhuff commented on Firestore/CHANGELOG.md:
```diff
@@ -2,6 +2,16 @@
+  currently supporting up to microsecond precision. It can be passed to API
+  methods anywhere a system Date is currently accepted. To make
+  `DocumentSnapshot`s read timestamp fields back as `Timestamp`s instead of
+  Dates, you can set the newly-added flag `areTimestampsInSnapshotsEnabled` in
```
I believe the setter is actually called `setTimestampsInSnapshotsEnabled`. It's exposed as a property though so it's also accessible as just `timestampsInSnapshotsEnabled`.

Therefore I think you want this phrase to be

To make ..., you can set the newly added `timestampsInSnapshotsEnabled` property in `FirestoreSettings` to `true`.

Please add a swift compilation test to verify that this is indeed true.

var-const replied:
Hmm, I don't know anything about Swift, but this doesn't seem to work:
* `timestampsInSnapshotsEnabled` gives _"'timestampsInSnapshotsEnabled' has been renamed to 'areTimestampsInSnapshotsEnabled'. 'timestampsInSnapshotsEnabled' was obsoleted in Swift 3"_;
* `setTimestampsInSnapshotsEnabled` gives _"Value of type `FirestoreSettings` has no member `setTimestampsInSnapshots`.

`areTimestampsInSnapshotsEnabled` is the only one that seems to work (note that the existing Swift build test also has `settings.isPersistenceEnabled = true`). Added this line to Swift compilation test (and also instantiation of `Timestamp` in `types` function). Removed the hyphen in "newly-added" and replaced "flag" with "property" in changelog entry.

wilhuff replied:
Ah right. Sorry about that. In Objective-C and early versions of Swift the property could have a different getter and setter name. The test shows you're definitely right. Thanks!
